File: ExcelDivider/src/main/java/ru/tsystems/divider/service/impl/RowToEntityConverterImpl.java
package ru.tsystems.divider.service.impl;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import org.apache.log4j.Logger;
import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.Row;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import ru.tsystems.divider.components.api.MessageWorker;
import ru.tsystems.divider.entity.Comment;
import ru.tsystems.divider.entity.Task;
import ru.tsystems.divider.service.api.*;

/**
 * It should be rebuilded, in my opinion. All file readings move to a ExcelFileReader. There should be only indexes of
 * row and cell. Or not.
 */
@Service
public class RowToEntityConverterImpl implements RowToEntityConverter {
    private static final Logger logger = Logger.getLogger(RowToEntityConverterImpl.class);

    private final Map<String, Integer> FIELDS;

    private final String KEY_MODIFICATOR;

    private static final String SOURCE = "column.index.";

    /**
     * Oh my lord. After add properties and delete hardcode, complete time upped 5-10 times. Wow, just Wow. Maybe some
     * clever people can do smthng better then this solution, which looks like crutch.
     */
    public RowToEntityConverterImpl(@Autowired MessageWorker messageWorker) {

        KEY_MODIFICATOR = messageWorker.getSourceValue("modificator.keys.pre");

        FIELDS = new HashMap<>();
        FIELDS.put("keys", getIndex("keys", messageWorker));
        FIELDS.put("summary", getIndex("summary", messageWorker));
        FIELDS.put("issueType", getIndex("issueType", messageWorker));
        FIELDS.put("status", getIndex("status", messageWorker));
        FIELDS.put("priority", getIndex("priority", messageWorker));
        FIELDS.put("resolution", getIndex("resolution", messageWorker));
        FIELDS.put("assignee", getIndex("assignee", messageWorker));
        FIELDS.put("reporter", getIndex("reporter", messageWorker));
        FIELDS.put("creater", getIndex("creater", messageWorker));
        FIELDS.put("created", getIndex("created", messageWorker));
        FIELDS.put("lastViewed", getIndex("lastViewed", messageWorker));
        FIELDS.put("updated", getIndex("updated", messageWorker));
        FIELDS.put("resolved", getIndex("resolved", messageWorker));
        FIELDS.put("affectsVersion", getIndex("affectsVersion", messageWorker));
        FIELDS.put("fixVersion", getIndex("fixVersion", messageWorker));
        FIELDS.put("components", getIndex("components", messageWorker));
        FIELDS.put("dueDate", getIndex("dueDate", messageWorker));
        FIELDS.put("originalEstimate", getIndex("originalEstimate", messageWorker));
        FIELDS.put("remainingEstimate", getIndex("remainingEstimate", messageWorker));
        FIELDS.put("timeSpent", getIndex("timeSpent", messageWorker));
        FIELDS.put("workRatio", getIndex("workRatio", messageWorker));
        FIELDS.put("description", getIndex("description", messageWorker));
        FIELDS.put("progress", getIndex("progress", messageWorker));
        FIELDS.put("sumProgress", getIndex("sumProgress", messageWorker));
        FIELDS.put("sumTimeSpent", getIndex("sumTimeSpent", messageWorker));
        FIELDS.put("sumRemainingEstimate", getIndex("sumRemainingEstimate", messageWorker));
        FIELDS.put("sumOriginalEstimate", getIndex("sumOriginalEstimate", messageWorker));
        FIELDS.put("labels", getIndex("labels", messageWorker));
        FIELDS.put("teams", getIndex("teams", messageWorker));
        FIELDS.put("epicName", getIndex("epicName", messageWorker));
        FIELDS.put("epicStatus", getIndex("epicStatus", messageWorker));
        FIELDS.put("epicColor", getIndex("epicColor", messageWorker));
        FIELDS.put("sprint", getIndex("sprint", messageWorker));
        FIELDS.put("epicLink", getIndex("epicLink", messageWorker));
        FIELDS.put("orderNumber", getIndex("orderNumber", messageWorker));
        FIELDS.put("deliveredVersion", getIndex("deliveredVersion", messageWorker));
        FIELDS.put("drcNumber", getIndex("drcNumber", messageWorker));
        FIELDS.put("keyword", getIndex("keyword", messageWorker));
        FIELDS.put("fixPriority", getIndex("fixPriority", messageWorker));
        FIELDS.put("commentStart", getIndex("commentStart", messageWorker));
        FIELDS.put("subTasks", getIndex("subTasks", messageWorker));
        FIELDS.put("relationTasks", getIndex("relationTasks", messageWorker));
        FIELDS.put("duplicateTasks", getIndex("duplicateTasks", messageWorker));
        FIELDS.put("commentMode.commentStart", getIndex("commentMode.commentStart", messageWorker));

    }

    @Autowired
    private EntityBuilder builder;

    @Autowired
    private FieldBuilder fieldBuilder;

    @Autowired
    private TaskService taskService;

    @Autowired
    private CommentService commentService;

    /**
     * Convert row to Entities. Save all entities into DB.
     *
     * @param row
     *            Row which should be converted.
     */
    @Override
    @Transactional
    public void addTaskFromRow(Row row) {
        String key = getStringCellValue(getCell(row, "keys"));
        if (KEY_MODIFICATOR != null)
            key = fieldBuilder.buildTaskKey(key, KEY_MODIFICATOR);
        if (taskService.findByKey(key) == null) {
            this.createNewTask(row);
        }
    }

    private void createNewTask(Row row) {
        Task task = new Task();
        task.setKeys(fieldBuilder.buildTaskKey(getStringCellValue(getCell(row, "keys")), KEY_MODIFICATOR)); // 1
        task.setSummary(getStringCellValue(getCell(row, "summary")));
        task.setIssueType(builder.buildIssueType(getStringCellValue(getCell(row, "issueType"))));
        task.setStatus(builder.buildStatus(getStringCellValue(getCell(row, "status"))));
        task.setPriority(builder.buildPriority(getStringCellValue(getCell(row, "priority"))));// 5
        task.setResolution(builder.buildResolution(getStringCellValue(getCell(row, "resolution"))));
        task.setAssignee(builder.buildEmployee(getStringCellValue(getCell(row, "assignee"))));
        task.setReporter(builder.buildEmployee(getStringCellValue(getCell(row, "reporter"))));
        task.setCreater(builder.buildEmployee(getStringCellValue(getCell(row, "creater"))));
        task.setCreated(getDateCellValue(getCell(row, "created")));// 10
        task.setLastViewed(getDateCellValue(getCell(row, "lastViewed")));// 11
        task.setUpdated(getDateCellValue(getCell(row, "updated")));
        task.setResolved(getDateCellValue(getCell(row, "resolved")));
        task.setAffectsVersions(builder.buildVersions(getStringCellValue(getCell(row, "affectsVersion"))));
        task.setFixVersions(builder.buildVersions(getStringCellValue(getCell(row, "fixVersion"))));// 15
        task.setComponents(builder.buildComponents(getStringCellValue(getCell(row, "components"))));// 16
        task.setDueDate(getDateCellValue(getCell(row, "dueDate")));
        task.setOriginalEstimate(getNumericCellValue(getCell(row, "originalEstimate")));
        task.setRemainingEstimate(getNumericCellValue(getCell(row, "remainingEstimate")));
        task.setTimeSpent(getNumericCellValue(getCell(row, "timeSpent")));// 20
        task.setWorkRation(getPercentCellValue(getCell(row, "workRatio"))); // percent value => need to multiply to 100
        task.setDescription(getStringCellValue(getCell(row, "description")));
        task.setProgress(getPercentCellValue(getCell(row, "progress"))); // percent value => need to multiply to 100
        task.setSumProgress(getPercentCellValue(getCell(row, "sumProgress"))); // percent value => need to multiply to
                                                                               // 100
        task.setSumTimeSpant(getNumericCellValue(getCell(row, "sumTimeSpent")));
        task.setSumRemainingEstimate(getNumericCellValue(getCell(row, "sumRemainingEstimate")));
        task.setSumOriginalEstimate(getNumericCellValue(getCell(row, "sumOriginalEstimate")));
        task.setLabels(builder.buildLabels(getStringCellValue(getCell(row, "labels"))));
        task.setTeams(builder.buildTeams(getStringCellValue(getCell(row, "teams"))));// 30
        task.setEpicName(getStringCellValue(getCell(row, "epicName")));
        task.setEpicStatus(builder.buildStatus(getStringCellValue(getCell(row, "epicStatus"))));
        task.setEpicColor(builder.buildEpicColor(getStringCellValue(getCell(row, "epicColor"))));
        task.setSprint(builder.buildSprint(getStringCellValue(getCell(row, "sprint"))));
        task.setEpicLink(getStringCellValue(getCell(row, "epicLink")));// 35
        task.setOrderNumber(getStringCellValue(getCell(row, "orderNumber")));
        task.setDeliveredVersion(getStringCellValue(getCell(row, "deliveredVersion")));
        task.setDrcNumber(getStringCellValue(getCell(row, "drcNumber")));
        task.setKeyword(builder.buildKeyword(getStringCellValue(getCell(row, "keyword"))));
        task.setFixPriority(builder.buildPriority(getStringCellValue(getCell(row, "fixPriority"))));// 40

        taskService.persist(task);

        saveAllComments(row, task);
    }

    @Override
    @Transactional
    public void addTasksConnectFromRow(Row row) {
        String key = fieldBuilder.buildTaskKey(getStringCellValue(getCell(row, "keys")), KEY_MODIFICATOR);
        Task task = taskService.findByKey(key);

        if (task == null)
            throw new IllegalArgumentException();

        task.setSubTasks(builder.buildConnectionToAnotherTasks(getStringCellValue(getCell(row, "subTasks"))));
        task.setRelationTasks(builder.buildConnectionToAnotherTasks(getStringCellValue(getCell(row, "relationTasks"))));
        task.setDuplicateTasks(builder
                .buildConnectionToAnotherTasks(getStringCellValue(getCell(row, "duplicateTasks"))));
        taskService.merge(task);
    }

    @Override
    public void saveAllComments(Row row, Task task) {
        int currentIndex = FIELDS.get("commentStart");
        String commentString;
        Comment curComment;

        while ((commentString = getStringCellValue(row.getCell(currentIndex))) != null) {
            curComment = builder.buildComments(commentString);
            curComment.setTask(task);
            commentService.persist(curComment);
            currentIndex++;
        }
    }

    @Override
    @Transactional
    public void saveAllComments(Row row) {
        int currentIndex = FIELDS.get("commentMode.commentStart");
        String commentString;
        Comment curComment;

        while ((commentString = getStringCellValue(row.getCell(currentIndex))) != null) {
            curComment = builder.buildCommentsWithTask(commentString);
            if (curComment != null){
                commentService.persist(curComment);
            }
            currentIndex++;
        }
    }

    private String getStringCellValue(Cell cell) {
        if (cell == null)
            return null;

        if (cell.getCellType() == Cell.CELL_TYPE_NUMERIC)
            return String.valueOf(cell.getNumericCellValue());
        if (cell.getCellType() == Cell.CELL_TYPE_ERROR)
            return String.valueOf(cell.getErrorCellValue()); // wow
        return cell.getStringCellValue();
    }

    private Date getDateCellValue(Cell cell) {
        if (cell == null)
            return null;
        return cell.getDateCellValue();
    }

    private Integer getNumericCellValue(Cell cell) {
        if (cell == null)
            return null;
        return doubleToInt(cell.getNumericCellValue());
    }

    private Integer getPercentCellValue(Cell cell) {
        if (cell == null)
            return null;
        return doubleToInt(cell.getNumericCellValue()) * 100;
    }

    private int doubleToInt(double value) {
        if (value < 1 && value > 0)
            value *= 100;
        return (int) value;
    }

    private static int getIndex(String fieldName, MessageWorker messageWorker) {
        try {
            String index = messageWorker.getSourceValue(SOURCE, fieldName);
            if (index == null)
                return -1;
            return Integer.valueOf(index);
        } catch (NumberFormatException nfexc) {
            logger.error("Wrong number: " + fieldName + "; Returned -1");
            return -1;
        }
    }

    private Cell getCell(Row row, String fieldName) {
        int index = FIELDS.get(fieldName);
        return index < 0 ? null : row.getCell(index);
    }
}

File: ExcelDivider/src/main/kt/ru/tsystems/divider/service/impl/RowToEntityConverterImpl.kt
package ru.tsystems.divider.service.impl

import org.apache.log4j.Logger
import org.apache.poi.ss.usermodel.Cell
import org.apache.poi.ss.usermodel.Row
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.tsystems.divider.components.api.MessageWorker
import ru.tsystems.divider.dao.api.CommentDao
import ru.tsystems.divider.dao.api.TaskDao
import ru.tsystems.divider.entity.Comment
import ru.tsystems.divider.entity.Consumables
import ru.tsystems.divider.entity.Dates
import ru.tsystems.divider.entity.Employee
import ru.tsystems.divider.entity.Epics
import ru.tsystems.divider.entity.Feature
import ru.tsystems.divider.entity.Statistics
import ru.tsystems.divider.entity.Task
import ru.tsystems.divider.entity.Workers
import ru.tsystems.divider.service.api.EntityBuilder
import ru.tsystems.divider.service.api.FieldBuilder
import ru.tsystems.divider.service.api.RowToEntityConverter
import ru.tsystems.divider.utils.constants.COMPONENT
import ru.tsystems.divider.utils.constants.DEFAULT
import ru.tsystems.divider.utils.constants.EPIC_COLOR
import ru.tsystems.divider.utils.constants.EPIC_STATUS
import ru.tsystems.divider.utils.constants.ISSUE_TYPE
import ru.tsystems.divider.utils.constants.KEYWORD
import ru.tsystems.divider.utils.constants.LABEL
import ru.tsystems.divider.utils.constants.PRIORITY
import ru.tsystems.divider.utils.constants.RESOLUTION
import ru.tsystems.divider.utils.constants.SPRINT
import ru.tsystems.divider.utils.constants.STATUS
import ru.tsystems.divider.utils.constants.TEAM
import ru.tsystems.divider.utils.constants.VERSION
import java.util.*
import kotlin.collections.HashSet

@Service
open class RowToEntityConverterImpl(
        @Autowired messageWorker: MessageWorker,
        @Autowired private val builder: EntityBuilder,
        @Autowired private val fieldBuilder: FieldBuilder,
        @Autowired private val taskDao: TaskDao,
        @Autowired private val commentDao: CommentDao
) : RowToEntityConverter {
    private val logger = Logger.getLogger(RowToEntityConverterImpl::class.java)

    private val FIELDS: MutableMap<String, Int> //todo: to map

    private val KEY_MODIFICATOR: String?

    private val SOURCE = "column.index."

    init {

        KEY_MODIFICATOR = messageWorker.getSourceValue("modificator.keys.pre")

        val getIndex: (fieldName: String) -> Int =
                { fieldName ->
                    try {
                        Integer.valueOf(messageWorker.getSourceValue(SOURCE, fieldName) ?: "-1")
                    } catch (nfexc: NumberFormatException) {
                        logger.error("Wrong number: $fieldName; Returned -1")
                        -1
                    }
                }

        FIELDS = HashMap()
        FIELDS["keys"] = getIndex("keys")
        FIELDS["summary"] = getIndex("summary")
        FIELDS["issueType"] = getIndex("issueType")
        FIELDS["status"] = getIndex("status")
        FIELDS["priority"] = getIndex("priority")
        FIELDS["resolution"] = getIndex("resolution")
        FIELDS["assignee"] = getIndex("assignee")
        FIELDS["reporter"] = getIndex("reporter")
        FIELDS["creater"] = getIndex("creater")
        FIELDS["created"] = getIndex("created")
        FIELDS["lastViewed"] = getIndex("lastViewed")
        FIELDS["updated"] = getIndex("updated")
        FIELDS["resolved"] = getIndex("resolved")
        FIELDS["affectsVersion"] = getIndex("affectsVersion")
        FIELDS["fixVersion"] = getIndex("fixVersion")
        FIELDS["components"] = getIndex("components")
        FIELDS["dueDate"] = getIndex("dueDate")
        FIELDS["originalEstimate"] = getIndex("originalEstimate")
        FIELDS["remainingEstimate"] = getIndex("remainingEstimate")
        FIELDS["timeSpent"] = getIndex("timeSpent")
        FIELDS["workRatio"] = getIndex("workRatio")
        FIELDS["description"] = getIndex("description")
        FIELDS["progress"] = getIndex("progress")
        FIELDS["sumProgress"] = getIndex("sumProgress")
        FIELDS["sumTimeSpent"] = getIndex("sumTimeSpent")
        FIELDS["sumRemainingEstimate"] = getIndex("sumRemainingEstimate")
        FIELDS["sumOriginalEstimate"] = getIndex("sumOriginalEstimate")
        FIELDS["labels"] = getIndex("labels")
        FIELDS["teams"] = getIndex("teams")
        FIELDS["epicName"] = getIndex("epicName")
        FIELDS["epicStatus"] = getIndex("epicStatus")
        FIELDS["epicColor"] = getIndex("epicColor")
        FIELDS["sprint"] = getIndex("sprint")
        FIELDS["epicLink"] = getIndex("epicLink")
        FIELDS["orderNumber"] = getIndex("orderNumber")
        FIELDS["deliveredVersion"] = getIndex("deliveredVersion")
        FIELDS["drcNumber"] = getIndex("drcNumber")
        FIELDS["keyword"] = getIndex("keyword")
        FIELDS["fixPriority"] = getIndex("fixPriority")
        FIELDS["commentStart"] = getIndex("commentStart")
        FIELDS["subTasks"] = getIndex("subTasks")
        FIELDS["relationTasks"] = getIndex("relationTasks")
        FIELDS["duplicateTasks"] = getIndex("duplicateTasks")
        FIELDS["commentMode.commentStart"] = getIndex("commentMode.commentStart")
    }


    /**
     * Convert row to Entities. Save all entities into DB.
     *
     * @param row
     * Row which should be converted.
     */
    @Transactional
    override fun addTaskFromRow(row: Row) {
        var key = getStringCellValue(getCell(row, "keys"))
        if (KEY_MODIFICATOR != null)
            key = fieldBuilder.buildTaskKey(key ?: throw java.lang.IllegalArgumentException("Wrong key for sub-task"),
                    KEY_MODIFICATOR)
        if (taskDao.getBykey(key) == null) {
            this.createNewTask(row)
        }
    }

    private fun createNewTask(row: Row) {
        val task = setUpTask(row)
        taskDao.persist(task)
        saveAllComments(row, task)
    }

    private fun setUpTask(row: Row): Task = Task(
            keys = fieldBuilder.buildTaskKey(getStringCellValue(getCell(row, "keys")
                    // todo: default key (DK) to constants or smthng else
            ) ?: throw IllegalArgumentException("Task's key cannot be null"), KEY_MODIFICATOR ?: "DK"),
            summary = getStringCellValue(getCell(row, "summary")),
            issueType = buildFeatureOrNull(getStringCellValue(getCell(row, "issueType")), ISSUE_TYPE),
            created = getDateCellValue(getCell(row, "created")),
            consumables = setUpConsumables(row)
    )

    private fun setUpConsumables(row: Row): Consumables = Consumables(
            affectsVersions = buildFeatureSetOrNull(getStringCellValue(getCell(row, "affectsVersion")), VERSION),
            components = buildFeatureSetOrNull(getStringCellValue(getCell(row, "components")), COMPONENT),
            deliveredVersion = buildFeatureOrNull(getStringCellValue(getCell(row, "deliveredVersion")), VERSION),
            description = getStringCellValue(getCell(row, "description")),
            drcNumber = getStringCellValue(getCell(row, "drcNumber")),
            fixPriority = buildFeatureOrNull(getStringCellValue(getCell(row, "fixPriority")), PRIORITY),
            fixVersions = buildFeatureSetOrNull(getStringCellValue(getCell(row, "fixVersion")), VERSION),
            keyword = buildFeatureOrNull(getStringCellValue(getCell(row, "keyword")), KEYWORD),
            labels = buildFeatureSetOrNull(getStringCellValue(getCell(row, "labels")), LABEL),
            orderNumber = getStringCellValue(getCell(row, "orderNumber")),
            priority = buildFeatureOrNull(getStringCellValue(getCell(row, "priority")), PRIORITY),
            resolution = buildFeatureOrNull(getStringCellValue(getCell(row, "resolution")), RESOLUTION),
            sprint = buildFeatureOrNull(getStringCellValue(getCell(row, "sprint")), SPRINT),
            status = buildFeatureOrNull(getStringCellValue(getCell(row, "status")), STATUS),
            teams = buildFeatureSetOrNull(getStringCellValue(getCell(row, "teams")), TEAM),

            dates = setUpDates(row),
            epics = setUpEpics(row),
            statistics = setUpStatistics(row),
            workers = setUpWorkers(row)
    )

    private fun setUpWorkers(row: Row): Workers = Workers(
            assignee = buildEmployeeOrNull(getStringCellValue(getCell(row, "assignee"))),
            creater = buildEmployeeOrNull(getStringCellValue(getCell(row, "creater"))),
            reporter = buildEmployeeOrNull(getStringCellValue(getCell(row, "reporter")))
    )

    private fun setUpEpics(row: Row): Epics = Epics(
            epicColor = buildFeatureOrNull(getStringCellValue(getCell(row, "epicColor")), EPIC_COLOR),
            epicLink = getStringCellValue(getCell(row, "epicLink")),
            epicName = getStringCellValue(getCell(row, "epicName")),
            epicStatus = buildFeatureOrNull(getStringCellValue(getCell(row, "epicStatus")), EPIC_STATUS)
    )

    private fun setUpStatistics(row: Row): Statistics = Statistics(
            original_estimate = getNumericCellValue(getCell(row, "originalEstimate")),
            progress = getPercentCellValue(getCell(row, "progress")),
            remaining_estimate = getNumericCellValue(getCell(row, "remainingEstimate")),
            time_spent = getNumericCellValue(getCell(row, "timeSpent")),
            work_ration = getPercentCellValue(getCell(row, "workRatio")),
            sum_progress = getPercentCellValue(getCell(row, "sumProgress")),
            sum_time_spant = getNumericCellValue(getCell(row, "sumTimeSpent")),
            sum_remaining_estimate = getNumericCellValue(getCell(row, "sumRemainingEstimate")),
            sum_original_estimate = getNumericCellValue(getCell(row, "sumOriginalEstimate"))
    )

    private fun setUpDates(row: Row): Dates = Dates(
            lastViewed = getDateCellValue(getCell(row, "lastViewed")),
            updated = getDateCellValue(getCell(row, "updated")),
            resolved = getDateCellValue(getCell(row, "resolved")),
            dueDate = getDateCellValue(getCell(row, "dueDate"))
    )

    private fun buildEmployeeOrNull(employee: String?):
            Employee? = if(employee == null) null else builder.buildEmployee(employee)

    private fun buildFeatureOrNull(title: String?, nature: String = DEFAULT):
            Feature? = if(title == null) null else builder.buildFeature(title, nature)

    private fun buildFeatureSetOrNull(title: String?, nature: String = DEFAULT):
            Set<Feature> = if(title == null) HashSet() else builder.buildFeatureSet(title, nature)

    private fun buildTaskConnection(task: String?) :
            Set<Task> = if(task == null) HashSet() else builder.buildConnectionToAnotherTasks(task)


    @Transactional
    override fun addTasksConnectFromRow(row: Row) {
        val key = fieldBuilder.buildTaskKey(getStringCellValue(getCell(row, "keys"))
        // todo: default key (DK) to constants or smthng else
                ?: throw IllegalArgumentException("Task's key cannot be null"), KEY_MODIFICATOR ?: "DK")
        val task = taskDao.getBykey(key) ?: throw IllegalArgumentException()

        task.subTasks = buildTaskConnection(getStringCellValue(getCell(row, "subTasks")))
        task.relationTasks = buildTaskConnection(getStringCellValue(getCell(row, "relationTasks")))
        task.duplicateTasks = buildTaskConnection(getStringCellValue(getCell(row, "duplicateTasks")))

        taskDao.merge(task)
    }

    @Transactional
    override fun saveAllComments(row: Row, task: Task) {
        var currentIndex = FIELDS["commentStart"]
                ?: throw java.lang.IllegalArgumentException("commentStart cannot to be null!")
        var commentString: String? = null
        var curComment: Comment?

        //todo : delete duplicates
        while ({ commentString = getStringCellValue(row.getCell(currentIndex++)); commentString }() != null) {
            curComment = builder.buildComments(commentString!!) //todo
            curComment!!.task = task
            commentDao.persist(curComment)
        }
    }


    @Transactional
    override fun saveAllComments(row: Row) { //todo: rename or delete method
        var currentIndex: Int = FIELDS["commentMode.commentStart"]
                ?: throw java.lang.IllegalArgumentException("commentMode.commentStart cannot to be null!")
        var commentString: String? = null
        var curComment: Comment?

        while ({ commentString = getStringCellValue(row.getCell(currentIndex++)); commentString }() != null) {
            curComment = builder.buildCommentsWithTask(commentString!!) //todo
            if (curComment != null) {
                commentDao.persist(curComment)
            }
        }
    }

    //todo : methods to EXCEL READER
    private fun getStringCellValue(cell: Cell?): String? {
        if (cell == null)
            return null

        if (cell.cellType == Cell.CELL_TYPE_NUMERIC)
            return cell.numericCellValue.toString()
        return if (cell.cellType == Cell.CELL_TYPE_ERROR) cell.errorCellValue.toString() else cell.stringCellValue // wow
    }

    private fun getDateCellValue(cell: Cell?): Date? {
        return cell?.dateCellValue
    }

    private fun getNumericCellValue(cell: Cell?): Int? {
        return if (cell == null) null else doubleToInt(cell.numericCellValue)
    }

    private fun getPercentCellValue(cell: Cell?): Int? {
        return if (cell == null) null else doubleToInt(cell.numericCellValue) * 100
    }
    //todo : methods to EXCEL READER

    private fun doubleToInt(value: Double): Int { //todo: smthng strange
        var value = value
        if (value < 1 && value > 0)
            value *= 100.0
        return value.toInt()
    }

    //todo: not right.. mb get value?
    private fun getCell(row: Row, fieldName: String): Cell? {
        val index: Int = FIELDS[fieldName] ?: -1
        return if (index < 0) null else row.getCell(index)
    }
}